<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>FastPCA with Single Cell Spatial Transcriptomics • FastPCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="FastPCA with Single Cell Spatial Transcriptomics">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">FastPCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.2.0000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro.html">Using FastPCA on Large Matrices</a>
    </li>
    <li>
      <a href="../articles/SpatialTranscriptomics.html">FastPCA with Single Cell Spatial Transcriptomics</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>FastPCA with Single Cell Spatial
Transcriptomics</h1>
            
      

      <div class="hidden name"><code>SpatialTranscriptomics.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">FastPCA</span><span class="op">)</span></span>
<span><span class="co">#&gt; Thank you for using FastPCA!</span></span>
<span><span class="co">#&gt; To cite this package, run: citation('FastPCA')</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="current-landscape-of-biological-data-analysis">Current Landscape of Biological Data Analysis<a class="anchor" aria-label="anchor" href="#current-landscape-of-biological-data-analysis"></a>
</h2>
<p>Data used for studying diseases is increasing at a fast pace. We used
to profile ~30,000 genes across 20 samples, then ~30,000 genes across
1,000 samples, now we can profile 18k genes across millions of cells
using Bruker (Nanostring) CosMx SMI. We can also image samples in high
dimensions with mass spec to get lipid and metabolite profiles from 5x5
um bins in a whole tissue section with ~10,000 unique peaks. Who knows
what the future will hold in terms of data dimensionality, but we need
to be able to maintain as much information from the whole data as
possible while also making it manageable to analyze. This is where
dimension reductions like PCA play a big role.</p>
<p>Currently, with how large data has gotten, the widely accepted way to
do this is to identify those features (genes, peaks, etc) that have the
largest variation between samples (cells, bins, pixels, etc), then use
only those to calculate the embedding. But can imagine that there is
information in those other features that are then being lost. The great
<code>irlba</code> package and method performs this dimension reduction
accurately but does take time and the R package doesn’t offer
multithreading which means it can take a very long time depending on how
many dimensions are being extracted and the number of samples which are
included.</p>
<p>This got me thinking “Is there another way to identify these top
dimensions that can be meaningfully multithreaded and produce results
that are almost as accurate for the sake of clustering samples?”. The
field of machine learning has long depended on matrix multiplication to
perform operations on large feature-space, leading to the logical path
of using their optimized code-bases to do the matrix operations on our
high dimensional biological data. <code>FastPCA</code> originally was
started because of my experience with <code>pytorch</code>. <a href="https://github.com/RafaelSdeSouza" class="external-link">Dr. Rafael S. de Souza</a>
created a package <a href="https://github.com/RafaelSdeSouza/qrpca" class="external-link"><code>qrpca</code></a> is
a similar thought to <code>FastPCA</code>, though using Rs <a href="https://cran.r-project.org/web/packages/torch/index.html" class="external-link"><code>torch</code></a>
impementation. However, <code>qrpca</code> doesn’t seem to produce
reduced-spaced singular value decomposition like <code>FastPCA</code>
does with <a href="https://epubs.siam.org/doi/10.1137/090771806" class="external-link">randomized SVD</a>.
Because my experience with <code>pytorch</code>, I wanted to do this
with python through <code>reticulate</code> rather than with Rs
<code>torch</code> (which actually ended up providing more benefits in
terms of speed, but does require careful usage because of system-level
conflicts). <strong>Due to CRANs checks, the defaults in the package are
using base R and <code>irlba</code>, but parameter selection can change
the backends.</strong></p>
<p>Here, we will walk through how to use <code>FastPCA</code> with the
<code>pytorch</code> backend, and discuss how this meaningfully differs
from both <code>irlba</code> and Dr. Rafael’s <code>qrpca</code>.</p>
</div>
<div class="section level2">
<h2 id="prepping-conda-environment">Prepping Conda Environment<a class="anchor" aria-label="anchor" href="#prepping-conda-environment"></a>
</h2>
<p>Within <code>FastPCA</code>, I’ve included a function called
<code><a href="../reference/setup_py_env.html">setup_py_env()</a></code> which makes the creation of the python
environment with <code>reticulate</code> easy to do. The recommended
method is <code>'conda'</code> since that’s what I have had the most
experience with and it allows for installing CUDA dependencies as I
demonstrated <a href="https://github.com/ACSoupir/cuda_conda" class="external-link">here</a>.
If wanting to use CUDA, I’d suggest creating the environment from
scratch to ensure that the pytorch install can see the CUDA device as
shown in the repo. Otherwise, it’s rather straightforward and can be ran
by calling:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/setup_py_env.html">setup_py_env</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"conda"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="starting-the-environment">Starting the Environment<a class="anchor" aria-label="anchor" href="#starting-the-environment"></a>
</h2>
<p>Once the environment is created, then it can be started. **If using
RStudio, I recommend restarting the session again just to be sure that
there isn’t anything that would cause conflicts. Again, inside of
<code>FastPCA</code> I included a function to start the environment:
<code><a href="../reference/start_FastPCA_env.html">start_FastPCA_env()</a></code>. If the defaults were used when setting
up the environment, the defaults can be used when starting the
environment as well.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/start_FastPCA_env.html">start_FastPCA_env</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Just as a sanity check to make sure that the python packages are
available, I typically check with <code>reticulate</code>. When using
packages like <code>Seurat</code> or <code>torch</code>,
<code>reticulate</code> sometimes loses track of where packages are even
though it can see the correct python version within the environment and
the config paths are all correct. Not sure how this can be fixed without
basically writing my own R-python handler that builds function calls and
then launches them on the terminal before reading it back to R, so be
mindful that certain orders can cause crashes and errors and that they
aren’t <code>FastPCA</code>’s fault but rather something deeper in the
software. One example is <code>reticulate</code> won’t be able to load
<code>pytorch</code> if R’s <code>torch</code> has already been used
somewhere in the same session.</p>
<p>Here we can see the config is correctly pointing to the conda
environment for <code>FastCPA</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/py_config.html" class="external-link">py_config</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; python:         /opt/anaconda3/envs/FastPCA/bin/python</span></span>
<span><span class="co">#&gt; libpython:      /opt/anaconda3/envs/FastPCA/lib/libpython3.10.dylib</span></span>
<span><span class="co">#&gt; pythonhome:     /opt/anaconda3/envs/FastPCA:/opt/anaconda3/envs/FastPCA</span></span>
<span><span class="co">#&gt; version:        3.10.18 | packaged by conda-forge | (main, Jun  4 2025, 14:46:00) [Clang 18.1.8 ]</span></span>
<span><span class="co">#&gt; numpy:          /opt/anaconda3/envs/FastPCA/lib/python3.10/site-packages/numpy</span></span>
<span><span class="co">#&gt; numpy_version:  2.2.6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; NOTE: Python version was forced by use_python(, required = FALSE)</span></span></code></pre></div>
<p>And here, can see that the package for <code>pytorch</code> is
available in the current environment.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/py_module_available.html" class="external-link">py_module_available</a></span><span class="op">(</span><span class="st">"torch"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<p>If this returns <code>FALSE</code>, likely either
<code><a href="../reference/start_FastPCA_env.html">start_FastPCA_env()</a></code> hasn’t been ran or, if the config
prints everything write but still get <code>FALSE</code> from this, then
likely need to restart the environment for reasons beyond my
knowledge.</p>
</div>
<div class="section level2">
<h2 id="spatial-transcriptomic-data">Spatial Transcriptomic Data<a class="anchor" aria-label="anchor" href="#spatial-transcriptomic-data"></a>
</h2>
<p>Previously, we performed single-cell spatial transcriptomics using
Nanostring (now Bruker) CosMx SMI to identify changes to the tumor
microenvironment and specific cell types associated with whether the
tumors had been exposed to immunotherapy or not (<span class="citation">Soupir et al. (2024)</span>). This is a large study in
terms of uniquely profiled tissues (tumor and stroma from ~20 patients)
and number of cells (~200,000) while the number of genes was small
(~1000). As mentioned at the beginning, there are now panels that are
18,000+ genes and can be applied across whole tissues resulting in
millions of cells. Because the focus of this vignette is on the
performance of the PCA, we are going to compare <code>FastPCA</code> to
<code>Seurat</code>’s. For larger feature-space and sample count,
<code>Seurat</code> will either:</p>
<ul>
<li>identify the most variables features, select them from the data and
perform PCA on those, or</li>
<li>create a ‘sketch’? of the high sample count (essentially down sample
to a smaller number of samples), identify most variable features, select
those features and perform PCA on the smaller sample x feature matrix,
then use the model from the smaller data set to project the full data in
lower dimension</li>
</ul>
<p>Both of these remove information form the data when performing
dimension reduction. One benefit is that it is more memory friendly.
However, those that should be helping in the analysis of this type of
data should have access to work stations and high performance computing
resources, both of which aren’t (shouldn’t be?) low memory systems. Even
with the <code>Seurat</code> object from this study, when imported into
R it’s 2.2GB in size. The data get’s large quickly (and this is with
sparse matrices).</p>
<p>The dataset is publically available on <a href="https://zenodo.org/doi/10.5281/zenodo.12730226" class="external-link">Zenodo</a> and can
be downloaded locally to be used. I have saved mine to my lab folder.
Here, it’s read into the environment and then we can start working with
it.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"/Volumes/lab_soupir/spatial_transcriptomics/example_data/seurat_object.Rds"</span><span class="op">)</span></span></code></pre></div>
<p>##Prepping the Expression Matrix</p>
<p>The results here will look different than from the above mentioned
manuscript because we are going to process differently (not
SCTransform). Will perform the normalization (not scaling) outside of
the <code>FastPCA</code> package but will use <code><a href="../reference/prep_matrix.html">prep_matrix()</a></code>
to transform and scale the data. The <code><a href="../reference/prep_matrix.html">prep_matrix()</a></code> function
will perform log2 transformation but sometimes <code><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p()</a></code> is
used instead.</p>
<p>The raw count data is in the object at
<code>seurat_obj@assays$Nanostring@counts</code> and in sparse format.
How <code>Seurat</code> performs normalization is with a scaling factor
to bring expression in line between samples after finding the proportion
of total counts each gene contributes. This is then used with the
<code><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p()</a></code> function mentioned.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scale.factor</span> <span class="op">=</span> <span class="fl">10000</span></span>
<span><span class="va">count_mat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">@</span><span class="va">assays</span><span class="op">$</span><span class="va">Nanostring</span><span class="op">@</span><span class="va">counts</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in asMethod(object): sparse-&gt;dense coercion: allocating vector of size</span></span>
<span><span class="co">#&gt; 1.5 GiB</span></span>
<span><span class="va">count_mat_norm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">count_mat</span><span class="op">)</span><span class="op">)</span> <span class="op">*</span> <span class="va">scale.factor</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>One large advantage of using <code>FastPCA</code> with they
python environment is the ability to change the number of CPU cores on
the fly, which the <code>rtorch</code> backend does not support once
it’s used (will see <code>libtorch</code> warning messages printed out
in the terminal)</strong>. Since our data is columns as samples and rows
as features, it needs to be transformed to have samples as rows and
features as columns. Further, for singular vector decomposition the
values should be mean centered (mean of 0) and unit variance (variance
of 1) scaled. To do this, we can run:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prepped_mat</span> <span class="op">=</span> <span class="fu"><a href="../reference/prep_matrix.html">prep_matrix</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">count_mat_norm</span>,</span>
<span>                          log2 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                          transpose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                          backend <span class="op">=</span> <span class="st">"pytorch"</span>,</span>
<span>                          cores <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                          device <span class="op">=</span> <span class="st">"CPU"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-fastpca">Running <code>FastPCA()</code><a class="anchor" aria-label="anchor" href="#running-fastpca"></a>
</h2>
<p>Next is actually running the singular value decomposition with
<code><a href="../reference/FastPCA.html">FastPCA()</a></code>. Depending on the backend, there are different
parameters that can be passed. The default backend is <code>'r'</code>
or <code>'irlba'</code> which all of the extra parameters with
<code>...</code> are passed to. The important parameters are
<code>k</code> for the number of dimensions to return, <code>p</code>
for the number of extra dimensions to use to more accurately capture
information in the <code>k</code> dimensions, and <code>q_iter</code>
for the number of power iterations. <strong>The best way to increase
accuracy in the tail end dimensions is to increase <code>p</code> rather
than inceasing <code>q_iter</code>.</strong> The oversampling
<code>p</code> will increase memory because it increases the size of the
matrix but if the variance doesn’t differ much in the higher dimensions,
it has limited benefit. The <code>q_iter</code> parameter will increase
accuracy of the top dimensions, which are typically fairly accurately
estimated anyway.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fastpca_runs</span> <span class="op">=</span> <span class="fu">bench</span><span class="fu">::</span><span class="fu"><a href="https://bench.r-lib.org/reference/mark.html" class="external-link">mark</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/FastPCA.html">FastPCA</a></span><span class="op">(</span>input_r_matrix <span class="op">=</span> <span class="va">prepped_mat</span>,</span>
<span>                      k <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                      p <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                      q_iter <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                      exact <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                      backend <span class="op">=</span> <span class="st">"pytorch"</span>,</span>
<span>                      cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  min_time <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  min_iterations <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>We can look at the time that it took to calculate the dimensions
here:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fastpca_runs</span><span class="op">$</span><span class="va">time</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;   2.587   2.589   2.589   2.626   2.667   2.699</span></span></code></pre></div>
<p>We will look at the results in the end when we have also run
<code><a href="https://rdrr.io/pkg/irlba/man/irlba.html" class="external-link">irlba::irlba</a></code> to compare.</p>
</div>
<div class="section level2">
<h2 id="running-with-seurat">Running with Seurat<a class="anchor" aria-label="anchor" href="#running-with-seurat"></a>
</h2>
<p>The original export from the machine produced a <code>Seurat</code>
object that was a V4 object. We can create a V5 object from the count
matrix and meta data since we don’t care about the spatial information
for now.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj2</span> <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_mat</span>,</span>
<span>                                 meta.data <span class="op">=</span> <span class="va">seurat_obj</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Data is of class matrix. Coercing to dgCMatrix.</span></span></code></pre></div>
<p>The <code><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">Seurat::RunPCA()</a></code> function requires data to be
normalized and scaled. <code>Seurat</code> makes this pretty straight
forward. Since we will use all genes present (actually 959 genes for
this data), will pass in all of the features to the
<code><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">Seurat::NormalizeData()</a></code> and
<code><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">Seurat::ScaleData()</a></code> functions.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj2</span> <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">seurat_obj2</span>,</span>
<span>                                    features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seurat_obj2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Normalizing layer: counts</span></span>
<span><span class="co">#&gt; Warning: The following arguments are not used: features</span></span>
<span><span class="va">seurat_obj2</span> <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">seurat_obj2</span>,</span>
<span>                                features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seurat_obj2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span></code></pre></div>
<p>Since <code>Seurat</code> uses <code>irlba</code> as well behind the
scenes, to match the same dimension search that we used with
<code>FastPCA</code> we</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_run</span> <span class="op">=</span> <span class="fu">bench</span><span class="fu">::</span><span class="fu"><a href="https://bench.r-lib.org/reference/mark.html" class="external-link">mark</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">seurat_obj2</span>, </span>
<span>                 reduction.name <span class="op">=</span> <span class="st">"pca"</span>, </span>
<span>                 npcs <span class="op">=</span> <span class="fl">100</span>, </span>
<span>                 work <span class="op">=</span> <span class="fl">110</span>,</span>
<span>                 features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">seurat_obj2</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  min_time <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  min_iterations <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Some expressions had a GC in every iteration; so filtering is</span></span>
<span><span class="co">#&gt; disabled.</span></span></code></pre></div>
<p>Again, lets look at the time:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">seurat_run</span><span class="op">$</span><span class="va">time</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;   497.3   497.3   497.3   497.3   497.3   497.3</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="comparing-fastpca-to-irlba">Comparing FastPCA to IRLBA<a class="anchor" aria-label="anchor" href="#comparing-fastpca-to-irlba"></a>
</h2>
<p><code>FastPCA</code> and <code>irlba</code> return the left and right
singular vectors as well as the diagonal singular values, but
<code>Seurat</code> returns <code>stdev</code> which from looking at
their <code>Seurat:::RunPCA.default()</code> is the singular values
divided by the square root of the number of samples just like:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fastpca_out</span> <span class="op">=</span> <span class="va">fastpca_runs</span><span class="op">$</span><span class="va">result</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">fastpca_stdev_vec</span> <span class="op">=</span> <span class="va">fastpca_out</span><span class="op">$</span><span class="va">S</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">prepped_mat</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can visualize these together with some <code>ggplot2</code>
code.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="va">seurat_stdev</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    stdev <span class="op">=</span> <span class="va">seurat_run</span><span class="op">$</span><span class="va">result</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">reductions</span><span class="op">$</span><span class="va">pca</span><span class="op">@</span><span class="va">stdev</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>dim <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                  method <span class="op">=</span> <span class="st">"SeuratIrlba"</span><span class="op">)</span></span>
<span><span class="va">fastpca_stdev</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  stdev <span class="op">=</span> <span class="va">fastpca_stdev_vec</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>dim <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                method <span class="op">=</span> <span class="st">"FastPCA"</span><span class="op">)</span></span>
<span><span class="va">plot_dat</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span></span>
<span>  <span class="va">seurat_stdev</span>,</span>
<span>  <span class="va">fastpca_stdev</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">#generate plot</span></span>
<span><span class="va">plot_dat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">dim</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dim</span>, y <span class="op">=</span> <span class="va">stdev</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpatialTranscriptomics_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<p>We can see that in the dimension up to ~10, they are almost
identical. When they flatten out (little difference in the variance
explained between PCs) it’s more difficult for FastPCA. However, let’s
look at the variance that those dimensions explain. This can be done by
squaring the singular values then dividing them by the variance of the
full prepped matrix <code>prepped_mat</code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#signif(fastpca_out$S^2 / sum(prepped_mat^2)*100, digits = 4)</span></span>
<span><span class="va">plot_dat</span> <span class="op">=</span> <span class="va">plot_dat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>S <span class="op">=</span> <span class="va">stdev</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">prepped_mat</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                  var_explained <span class="op">=</span> <span class="va">S</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">prepped_mat</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">plot_dat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">method</span> <span class="op">==</span> <span class="st">"FastPCA"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">#&gt;       stdev dim  method         S var_explained</span></span>
<span><span class="co">#&gt; 1  5.595580   1 FastPCA 2496.8579     3.2015003</span></span>
<span><span class="co">#&gt; 2  3.860744   2 FastPCA 1722.7398     1.5240712</span></span>
<span><span class="co">#&gt; 3  3.397633   3 FastPCA 1516.0906     1.1803647</span></span>
<span><span class="co">#&gt; 4  2.784533   4 FastPCA 1242.5136     0.7928083</span></span>
<span><span class="co">#&gt; 5  2.417531   5 FastPCA 1078.7500     0.5975957</span></span>
<span><span class="co">#&gt; 6  2.098351   6 FastPCA  936.3256     0.4502147</span></span>
<span><span class="co">#&gt; 7  2.012898   7 FastPCA  898.1945     0.4142921</span></span>
<span><span class="co">#&gt; 8  1.948158   8 FastPCA  869.3064     0.3880714</span></span>
<span><span class="co">#&gt; 9  1.817298   9 FastPCA  810.9141     0.3376880</span></span>
<span><span class="co">#&gt; 10 1.652378  10 FastPCA  737.3235     0.2791785</span></span>
<span><span class="co">#&gt; 11 1.567201  11 FastPCA  699.3161     0.2511383</span></span>
<span><span class="co">#&gt; 12 1.529472  12 FastPCA  682.4803     0.2391917</span></span>
<span><span class="co">#&gt; 13 1.508728  13 FastPCA  673.2243     0.2327477</span></span>
<span><span class="co">#&gt; 14 1.394348  14 FastPCA  622.1854     0.1987950</span></span>
<span><span class="co">#&gt; 15 1.377546  15 FastPCA  614.6882     0.1940330</span></span></code></pre></div>
<p>View same for <code>Seurat</code>/<code>irlba</code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_dat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">method</span> <span class="op">==</span> <span class="st">"SeuratIrlba"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">#&gt;       stdev dim      method         S var_explained</span></span>
<span><span class="co">#&gt; 1  5.595807   1 SeuratIrlba 2496.9590     3.2017596</span></span>
<span><span class="co">#&gt; 2  3.860760   2 SeuratIrlba 1722.7473     1.5240844</span></span>
<span><span class="co">#&gt; 3  3.397714   3 SeuratIrlba 1516.1266     1.1804208</span></span>
<span><span class="co">#&gt; 4  2.784830   4 SeuratIrlba 1242.6461     0.7929775</span></span>
<span><span class="co">#&gt; 5  2.418789   5 SeuratIrlba 1079.3112     0.5982177</span></span>
<span><span class="co">#&gt; 6  2.101672   6 SeuratIrlba  937.8074     0.4516408</span></span>
<span><span class="co">#&gt; 7  2.017853   7 SeuratIrlba  900.4055     0.4163343</span></span>
<span><span class="co">#&gt; 8  1.953772   8 SeuratIrlba  871.8115     0.3903112</span></span>
<span><span class="co">#&gt; 9  1.825876   9 SeuratIrlba  814.7418     0.3408834</span></span>
<span><span class="co">#&gt; 10 1.674047  10 SeuratIrlba  746.9928     0.2865488</span></span>
<span><span class="co">#&gt; 11 1.589484  11 SeuratIrlba  709.2592     0.2583306</span></span>
<span><span class="co">#&gt; 12 1.556328  12 SeuratIrlba  694.4643     0.2476657</span></span>
<span><span class="co">#&gt; 13 1.541150  13 SeuratIrlba  687.6916     0.2428585</span></span>
<span><span class="co">#&gt; 14 1.449866  14 SeuratIrlba  646.9587     0.2149409</span></span>
<span><span class="co">#&gt; 15 1.438523  15 SeuratIrlba  641.8974     0.2115910</span></span></code></pre></div>
<p>Here we see that the just how little variance is explained by these
slightly higher dimensions, keeping in mind that there are <em>at
most</em> 978 dimensions before we get back to the full matrix. Speaking
of, lets quickly run the exact SVD using <code>FastPCA</code>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fastpca_runs_exact</span> <span class="op">=</span> <span class="fu">bench</span><span class="fu">::</span><span class="fu"><a href="https://bench.r-lib.org/reference/mark.html" class="external-link">mark</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/FastPCA.html">FastPCA</a></span><span class="op">(</span>input_r_matrix <span class="op">=</span> <span class="va">prepped_mat</span>,</span>
<span>                      k <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                      p <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                      q_iter <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                      exact <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                      backend <span class="op">=</span> <span class="st">"pytorch"</span>,</span>
<span>                      cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  min_time <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  min_iterations <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Even with the full SVD (left) matrix, it still ran:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fastpca_runs_exact</span><span class="op">$</span><span class="va">time</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">#&gt;   10.02   10.09   10.11   10.13   10.18   10.28</span></span></code></pre></div>
<p>Can now add these exact values to the plotting data and see how they
compare, still taking <em>FAR</em> less time than
<code>Seurat</code>/<code>irlba</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_dat_all</span> <span class="op">=</span> <span class="va">plot_dat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"FastPCA Exact"</span>,</span>
<span>                             S <span class="op">=</span> <span class="va">fastpca_runs_exact</span><span class="op">$</span><span class="va">result</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">S</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                     <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>stdev <span class="op">=</span> <span class="va">S</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">prepped_mat</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                   var_explained <span class="op">=</span> <span class="va">S</span><span class="op">^</span><span class="fl">2</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">prepped_mat</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span> <span class="op">*</span> <span class="fl">100</span>,</span>
<span>                                   dim <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">plot_dat_all</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">dim</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dim</span>, y <span class="op">=</span> <span class="va">stdev</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span>, alpha <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpatialTranscriptomics_files/figure-html/unnamed-chunk-22-1.png" width="768"></p>
<p>Lastly, lets look at the time it took to run each of these:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">time_dat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>time <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">fastpca_runs</span><span class="op">$</span><span class="va">time</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                               <span class="va">fastpca_runs_exact</span><span class="op">$</span><span class="va">time</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                               <span class="va">seurat_run</span><span class="op">$</span><span class="va">time</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                      method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"FastPCA"</span>, <span class="va">fastpca_runs</span><span class="op">$</span><span class="va">n_itr</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                                 <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"FastPCA Exact"</span>, <span class="va">fastpca_runs_exact</span><span class="op">$</span><span class="va">n_itr</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                                 <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Seurat/Irlba"</span>, <span class="va">seurat_run</span><span class="op">$</span><span class="va">n_itr</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">time_dat</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">method</span>, y <span class="op">=</span> <span class="va">time</span>, color <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Method"</span>, y <span class="op">=</span> <span class="st">"Time (s)"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="SpatialTranscriptomics_files/figure-html/unnamed-chunk-23-1.png" width="768"></p>
<p>Even on a log scale(!), <code>FastPCA</code> looks significantly
faster (because it is).</p>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-Soupir2024-ww" class="csl-entry">
Soupir, Alex C, Mitchell T Hayes, Taylor C Peak, Oscar Ospina, Nicholas
H Chakiryan, Anders E Berglund, Paul A Stewart, et al. 2024.
<span>“Increased Spatial Coupling of Integrin and Collagen
<span>IV</span> in the Immunoresistant Clear-Cell Renal-Cell Carcinoma
Tumor Microenvironment.”</span> <em>Genome Biol.</em> 25 (1): 308.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Alex Soupir.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
